### PR TITLE
Multioutput backward formula: allow conditional guards against saving

### PR DESCRIPTION
**Pull request by nikitaved**
Multi-output backward formulas break the ability of autogen to decide which variables have to be stored in a graph.
This PR introduces a macro `wrap_opt_if` which could be used to hint autogen about variable interdependence.

For example, the following code is being generated for `_trilinear` with this modification:
```
at::Tensor _trilinear(c10::DispatchKeySet ks, const at::Tensor & i1, const at::Tensor & i2, const at::Tensor & i3, at::IntArrayRef expand1, at::IntArrayRef expand2, at::IntArrayRef expand3, at::IntArrayRef sumdim, int64_t unroll_dim) {
  auto& i1_ = unpack(i1, "i1", 0);
  auto& i2_ = unpack(i2, "i2", 1);
  auto& i3_ = unpack(i3, "i3", 2);
  [[maybe_unused]] auto _any_requires_grad = compute_requires_grad( i1, i2, i3 );

  [[maybe_unused]] auto _any_has_forward_grad_result = (isFwGradDefined(i1) || isFwGradDefined(i2) || isFwGradDefined(i3));
  std::shared_ptr<TrilinearBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<TrilinearBackward0>(new TrilinearBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( i1, i2, i3 ));
    grad_fn->expand1 = expand1.vec();
    grad_fn->expand2 = expand2.vec();
    grad_fn->expand3 = expand3.vec();
    if (grad_fn->should_compute_output(1) || grad_fn->should_compute_output(2)) {
      grad_fn->i1_ = SavedVariable(i1, false);
    }
    if (grad_fn->should_compute_output(0) || grad_fn->should_compute_output(2)) {
      grad_fn->i2_ = SavedVariable(i2, false);
    }
    if (grad_fn->should_compute_output(0) || grad_fn->should_compute_output(1)) {
      grad_fn->i3_ = SavedVariable(i3, false);
    }
    grad_fn->sumdim = sumdim.vec();
  }

```

with the following backward modifications:
```
 - name: _trilinear(Tensor i1, Tensor i2, Tensor i3, int[] expand1, int[] expand2, int[] expand3, int[] sumdim, int unroll_dim=1) -> Tensor
  - i1, i2, i3: _trilinear_backward(grad, i1, i2, i3, expand1, expand2, expand3, sumdim, grad_input_mask)
  + i1, i2, i3: "_trilinear_backward(grad,
  +             wrap_opt_if(i1, grad_input_mask[1] || grad_input_mask[2]),
  +             wrap_opt_if(i2, grad_input_mask[0] || grad_input_mask[2]),
  +             wrap_opt_if(i3, grad_input_mask[0] || grad_input_mask[1]),
  +             expand1, expand2, expand3, sumdim, grad_input_mask)"
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101978
* #103548
* #104042
* #103544
* __->__ #103750

cc @ezyang @albanD @zou3519 @gqchen @pearu @soulitzer @Lezcano @Varal7 @bhosmer @bdhirsh

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
I'll let @soulitzer review this one.
What I had in mind was more of adding a replacement entry here: https://github.com/pytorch/pytorch/blob/ae78e80123fbb3cec6b0d0ce9f96c7e0062da602/tools/autograd/load_derivatives.py#L758 but maybe that doesn't work?

**Comment by nikitaved** [comment]
@albanD , nope, that did not work... The substituted formula is then being parsed through `guard_for` which saves everything for multi-output formulas by default.

File: torch/csrc/autograd/FunctionsManual.h [hunk 1]
```diff
@@ -38,6 +38,11 @@ Tensor toNonOptFwGrad(const c10::optional<Tensor>& t);
 Tensor toNonOptPrimal(const c10::optional<Tensor>& t);
 Tensor toNonOptTensor(const c10::optional<Tensor>& t);
 
+inline c10::optional<Tensor> wrap_opt_if(const Tensor& t, const bool cond) {
+  using OptTensor = c10::optional<Tensor>;
+  return cond ? OptTensor(t) : static_cast<OptTensor>(c10::nullopt);
+}
+
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.h:45]: For context: this one is used in the next PR up in the stack for `sparse_sampled_addmm_backward`.

**Comment by nikitaved** [comment]
The tests are missing though, I will figure them out later.

**Comment by albanD** [comment]
>  nope, that did not work... The substituted formula is then being parsed by guard_for which saves everything for multi-output formulas by default.

Ho sad. The new formula should just contain `arg_opt` right? Then the `expr` provided will be used to generate what is given to the SavedVariable.
Similar to how `self->sym_sizes()` is replaced by `self_sym_sizes_opt` in the formula and the saving formula is `    grad_fn->bias_sym_sizes_opt = bias.has_value() ? c10::optional<c10::SymIntArrayRef>(bias->sym_sizes()) : c10::nullopt;` where the decision is made during the fw to save something or not.

**Comment by nikitaved** [comment]
Tried, did not work as well :( Hence had to dig deeper.

**Comment by nikitaved** [comment]
@albanD , right, but expressions use several variables, so I need to match and back-substitute into the expression. Back-substitution seems possible only in `res`, but this is not what we need, right?

File: torch/csrc/autograd/FunctionsManual.h (hunk 1 above)
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.h:45]: If codegen did its job correctly, we would always get an undefined tensor t here right? Maybe we can assert for that here.

**Comment by soulitzer** [comment]
Looks pretty good! For testing, maybe just apply this to the multi-output formulas we already have?
Could you also add the codegen output before/after to the PR description.

File: torch/csrc/autograd/FunctionsManual.h (hunk 1 above)
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.h:45]: No, not really, unfortunately. This code is being run at backward compute. But we can assert inside backward implementations for sure.

Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.h:45]: oh whoops, good point.

**Comment by albanD** [comment]
Sounds good! Thanks for looking into it in details.

**Comment by nikitaved** [comment]
@soulitzer , @albanD , I have updated the PR description and modified one function for demonstration purposes. `sampled_addmm_backward` is also modified and it works, but it is in a separate PR.

**Review by soulitzer — APPROVED** [review]
Nice!

**Comment by soulitzer** [comment]
Oh @nikitaved  we also have documentation about `grad_input_mask` in the huge comment at the beginning of `derivatives.yaml`, we should probably mention this new thing there.

**Comment by nikitaved** [comment]
@soulitzer , all right, will do!

**Comment by nikitaved** [comment]
@soulitzer , could you please have a look at the comment?

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -110,6 +110,25 @@
 #     destroy saved buffers if we know variables are not going to be retained,
 #     e.g., it is used by _cudnn_rnn
 #
+# In a gradient expression, the following functions are in scope:
```
Comment (soulitzer) [inline tools/autograd/derivatives.yaml:113]: we don't need this additional header here right? it kind of sounds like this is the only function in scope in the gradient expression.

**Review by soulitzer — COMMENTED** [review]
comment looks great, small nit

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (nikitaved) [reply tools/autograd/derivatives.yaml:113]: This is true, but the section above is about variables. I could move it there while replacing `variables` with `variables/functions`.

Comment (soulitzer) [reply tools/autograd/derivatives.yaml:113]: Ahh mb didn't realize the difference, combining sounds good to me!

**Comment by nikitaved** [comment]
@soulitzer , the nit is addressed. If there is anything else, please, let me know. Otherwise it is ready to go :)

**Review by soulitzer — APPROVED** [review]
